metric-publishers/emf-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/emf/MetricEmfConverter.java

```diff
@@ -0,0 +1,368 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.publishers.emf;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.metrics.MetricCategory;
+import software.amazon.awssdk.metrics.MetricCollection;
+import software.amazon.awssdk.metrics.MetricRecord;
+import software.amazon.awssdk.protocols.jsoncore.JsonWriter;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * # MetricEmfConverter
+
+ * Converts metrics into Amazon CloudWatch Embedded Metric Format (EMF).
+ * This internal class handles the transformation of various metric types
+ * into EMF-compatible format.
+
+ * ## Configuration
+ * The converter is initialized with an `EmfMetricConfiguration` that defines:
+ * - Metric categories to process
+ * - Other EMF-specific settings
+
+ * ## Implementation Notes
+ * - Boolean values are converted to numeric values (1.0 for true, 0.0 for false)
+ * - Null values are converted to 0.0
+ * - Very small numeric values (below 0.0001) are normalized to 0.0
+ * - Duration values are converted to milliseconds
+ *
+ * @since 2.0
+ * @see EmfMetricConfiguration
+ */
+@SdkInternalApi
+public class MetricEmfConverter {
+    private static final Logger logger = Logger.loggerFor("software.amazon.awssdk.metrics.publishers.emf");
+    private static final double ZERO_THRESHOLD = 0.0001;
+    private final List<String> realDimensionStrings = new ArrayList<>();
+    private final EmfMetricConfiguration config;
+    private final boolean metricCategoriesContainsAll;
+
+    // Constructor that takes the configuration
+    public MetricEmfConverter(EmfMetricConfiguration config) {
+        this.config = config;
+        this.metricCategoriesContainsAll = config.getMetricCategories().contains(MetricCategory.ALL);
+    }
+
+    /**
+     * Processes and normalizes metric values for EMF formatting.
+     * The method handles various input types and normalizes them according to EMF requirements:
+     * - `null` : 0.0
+     * - `Boolean` : 1.0 (true) or 0.0 (false)
+     * - `Duration` : milliseconds value
+     * - `Double`: normalized to 0.0 if below threshold
+     *
+     * @param metricValue The metric value to process
+     * @return The processed value in EMF-compatible format
+     */
+    private Object processValue(Object metricValue) {
+        if (metricValue == null) {
+            return 0.0;
+        }
+
+        // Changes boolean value from true/false to 1/0
+        if (metricValue instanceof Boolean) {
+            return metricValue.equals(true) ? 1.0 : 0.0;
+        }
+
+        // Changes duration value to a number of ms
```

Comment:
Same here, the code should be self explanatory ideally, and we should only add comments in places that truly needs context.

Could you check other comments in this PR as well?